test/reliability_manager.js

```diff
@@ -0,0 +1,212 @@
+                assert(!trxSpy.called);
+            });
+        });
+
+        describe("Known version, reliabilityFactor exceeds reliabilityThreshold", () => {
+            let trx;
+            before(async () => {
+                const guid = randomBuffer(consts.GUID_SIZE);
+                const hash = randomBuffer(consts.HASH_SIZE);
+                await rm.processTransaction(new StablePutTransaction(guid, hash)); // once
+                await rm.processTransaction(new StablePutTransaction(guid, hash)); // twice
+                trx = new StablePutTransaction(guid, hash);
+                assert(trx.isValid);
+                await rm.processTransaction(trx); // three times (last time)
+            });
+
+            it("should not change the reliabilityFactor, regardless of versionHash consistency", () => {
+                const entry = rm.getEntry(helpers.GUIDBufferToString(trx.guid), trx.hash.toString('hex'));
+                assert(entry);
+                assert.strictEqual(entry.factor, 2);
+                assert.strictEqual(entry.state, ReliabilityManager.reliabilityStates.Reliable);
+            });
+
+            it("should invalidate the transaction to prevent changes to the version", () => {
+                assert(!trx.isValid);
+            });
+        });
+
+        describe("Known version, inconsistent versionHash", () => {
+            let trx, spy;
+            before(async () => {
+                const guid = randomBuffer(consts.GUID_SIZE);
+                const hash = randomBuffer(consts.HASH_SIZE);
+                await rm.processTransaction(new UnstablePutTransaction(guid, hash)); // once
+
+                trx = await new UnstablePutTransaction(guid, hash);
+                spy = sinon.spy(trx, "writeFilesToPath");
+                assert(trx.isValid);
+                await rm.processTransaction(trx); // twice (last time)
+            });
+
+            it("should set the correct reliability state", () => {
+                const entry = rm.getEntry(helpers.GUIDBufferToString(trx.guid), trx.hash.toString('hex'));
+                assert(entry);
+                assert.strictEqual(entry.state, ReliabilityManager.reliabilityStates.Unreliable);
+            });
+
+            it("should invalidate the transaction", () => {
+                assert(!trx.isValid);
+            });
+
+            it("should tell the transaction to write unreliable files if saveUnreliableVersionArtifacts is true", async () => {
+                assert(spy.called);
+
+                // Test with saveUnreliableVersionArtifacts = false
+                const myRm = new ReliabilityManager(db, tmp.tmpNameSync(), {reliabilityThreshold: 2, saveUnreliableVersionArtifacts: false});
+                spy.resetHistory();
+                await myRm.processTransaction(trx);
+                assert(!spy.called);
+            });
+        });
+
+        describe("multiClient", () => {
```

Comment:
It looks like this is testing transactions from 3 different client addresses - "A", "B", and "".

Reply:
good catch - this test as written was actually masking the bug you found where the clientId was being set on a new entry.